Fix Google instance template creation for preemptible VM

erobert-c opened the pull request:
Hi there,

This commit should allow the creation of preemptible instance template in google compute engine. Currently, while trying to create such template, we get the following error:

> Error applying plan:
> 
> 1 error(s) occurred:
> 
> \* google_compute_instance_template.preentibletemplate-test: Error creating instance: googleapi: Error 400: Invalid value for field 'resource.properties.scheduling.preemptible': 'true'.  Scheduling must have preemptible be false when AutomaticRestart is true., invalid

Here is a simple template to reproduce the issue
```hcl
resource "google_compute_instance_template" "preentibletemplate-test" {
  name = "preemptible-template-test"
  machine_type = "n1-standard-2"
  scheduling {
    automatic_restart = false
    preemptible = true
  }
  disk {
    source_image = "${var.base_image}"
    boot = true
  }
  network_interface {
    network = "${var.subnet_id}"
  }
}
```

Tests results with `make test TEST=./builtin/providers/google/` are still good.

Cheers

phinze commented:
This LGTM - would like a double check from @lwander to be sure though.

lwander commented on builtin/providers/google/resource_compute_instance_template.go:
```diff
@@ -403,16 +404,24 @@ func resourceComputeInstanceTemplateCreate(d *schema.ResourceData, meta interfac
+			forceSendFieldsScheduling = append(forceSendFieldsScheduling, "Preemptible")
+			if vp == true {
+				instanceProperties.Scheduling.OnHostMaintenance = "TERMINATE"
+				forceSendFieldsScheduling = append(forceSendFieldsScheduling, "OnHostMaintenance")
```
You allocate the capacity for this slice as `3`, but you potentially add `4` items. This can cause runtime panic. I would recommend only placing `"OnHostMaintenance"` into `ForceSendFields` once, to avoid sending it twice.

lwander commented on builtin/providers/google/resource_compute_instance_template.go:
```diff
@@ -403,16 +404,24 @@ func resourceComputeInstanceTemplateCreate(d *schema.ResourceData, meta interfac
 		if vp, okp := _scheduling["preemptible"]; okp {
 			instanceProperties.Scheduling.Preemptible = vp.(bool)
+			forceSendFieldsScheduling = append(forceSendFieldsScheduling, "Preemptible")
+			if vp == true {
```
`if vp { ... }` is preferred.

lwander commented:
Good catch and fix! Thanks for doing this @erobert-c; once you address the comments it's ready to be merged!

erobert-c commented:
Hi,

Thanks for the review, I'll fix those most probably beginning of next week.

erobert-c commented:
Hi @lwander, 

I tried to integrate your feedbacks, squashed the commits and rebased on master. Don't hesitate to tell if you'd like me to change something else.

Cheers

lwander commented:
LGTM!
